File: Kotlin/bloom_filter_utils/BloomFilterTest.kt
/**
 * Bloom Filter 测试模块
 * 
 * 测试布隆过滤器的各项功能
 * 
 * @author AllToolkit Auto-Generator
 * @date 2026-04-13
 */

package bloom_filter_utils

import kotlin.system.measureNanoTime

/**
 * 测试结果类
 */
data class TestResult(
    val name: String,
    val passed: Boolean,
    val message: String = ""
)

/**
 * 测试运行器
 */
object BloomFilterTest {
    private val results = mutableListOf<TestResult>()
    
    fun runAllTests() {
        println("========================================")
        println("  Bloom Filter 测试套件")
        println("========================================\n")
        
        // 基础功能测试
        testBasicOperations()
        testMightContain()
        testFalsePositiveRate()
        testClearOperation()
        testSerialization()
        
        // 可扩展布隆过滤器测试
        testScalableBloomFilter()
        
        // 计数布隆过滤器测试
        testCountingBloomFilter()
        
        // 性能测试
        testPerformance()
        testMemoryEfficiency()
        
        // 打印结果
        printResults()
    }
    
    /**
     * 测试基础操作
     */
    private fun testBasicOperations() {
        val testName = "基础操作测试"
        try {
            val filter = BloomFilter<String>(1000, 0.01)
            
            // 测试初始状态
            assert(filter.isEmpty()) { "新创建的过滤器应该为空" }
            assert(filter.size() == 0) { "初始大小应为0" }
            
            // 添加元素
            filter.add("hello")
            assert(!filter.isEmpty()) { "添加元素后不应为空" }
            assert(filter.size() == 1) { "大小应为1" }
            
            // 批量添加
            filter.addAll(listOf("world", "kotlin", "bloom"))
            assert(filter.size() == 4) { "批量添加后大小应为4" }
            
            results.add(TestResult(testName, true))
        } catch (e: Exception) {
            results.add(TestResult(testName, false, e.message ?: "Unknown error"))
        }
    }
    
    /**
     * 测试元素检查
     */
    private fun testMightContain() {
        val testName = "元素检查测试"
        try {
            val filter = BloomFilter<Int>(100, 0.01)
            
            // 添加1-50
            for (i in 1..50) {
                filter.add(i)
            }
            
            // 检查已添加的元素
            var allContained = true
            for (i in 1..50) {
                if (!filter.mightContain(i)) {
                    allContained = false
                    break
                }
            }
            assert(allContained) { "已添加的元素应该全部能被找到" }
            
            // 检查未添加的元素（注意可能有假阳性）
            // definitelyNotContains 应该对大部分未添加元素返回 true
            var definitelyNotCount = 0
            for (i in 51..100) {
                if (filter.definitelyNotContains(i)) {
                    definitelyNotCount++
                }
            }
            
            // 由于假阳性概率很低，大部分未添加元素应该确实不存在
            assert(definitelyNotCount >= 45) { "大部分未添加元素应该确实不存在" }
            
            results.add(TestResult(testName, true))
        } catch (e: Exception) {
            results.add(TestResult(testName, false, e.message ?: "Unknown error"))
        }
    }
    
    /**
     * 测试假阳性率
     */
    private fun testFalsePositiveRate() {
        val testName = "假阳性率测试"
        try {
            val expectedInsertions = 10000
            val targetFpp = 0.03
            val filter = BloomFilter<String>(expectedInsertions, targetFpp)
            
            // 添加元素
            for (i in 1..expectedInsertions) {
                filter.add("item_$i")
            }
            
            // 测试假阳性
            val testCount = 10000
            var falsePositives = 0
            for (i in (expectedInsertions + 1)..(expectedInsertions + testCount)) {
                if (filter.mightContain("item_$i")) {
                    falsePositives++
                }
            }
            
            val actualFpp = falsePositives.toDouble() / testCount
            println("\n  假阳性率测试:")
            println("    目标假阳性率: ${String.format("%.4f", targetFpp)}")
            println("    实际假阳性率: ${String.format("%.4f", actualFpp)}")
            println("    估计假阳性率: ${String.format("%.4f", filter.estimateFpp())}")
            
            // 实际假阳性率应该接近目标值（允许一定误差）
            // 注意：由于随机性，实际FPP可能会有波动
            assert(actualFpp < targetFpp * 3) { "实际假阳性率不应过高" }
            
            results.add(TestResult(testName, true, "实际FPP=${String.format("%.4f", actualFpp)}"))
        } catch (e: Exception) {
            results.add(TestResult(testName, false, e.message ?: "Unknown error"))
        }
    }
    
    /**
     * 测试清空操作
     */
    private fun testClearOperation() {
        val testName = "清空操作测试"
        try {
            val filter = BloomFilter<String>(100, 0.01)
            
            filter.addAll(listOf("a", "b", "c", "d", "e"))
            assert(!filter.isEmpty()) { "添加元素后不应为空" }
            
            filter.clear()
            assert(filter.isEmpty()) { "清空后应为空" }
            assert(filter.size() == 0) { "清空后大小应为0" }
            
            // 清空后添加应该正常工作
            filter.add("new")
            assert(filter.mightContain("new")) { "清空后应该能添加新元素" }
            
            results.add(TestResult(testName, true))
        } catch (e: Exception) {
            results.add(TestResult(testName, false, e.message ?: "Unknown error"))
        }
    }
    
    /**
     * 测试序列化
     */
    private fun testSerialization() {
        val testName = "序列化测试"
        try {
            val originalFilter = BloomFilter<Int>(1000, 0.01)
            
            // 添加元素
            for (i in 1..500) {
                originalFilter.add(i)
            }
            
            // 序列化
            val bytes = originalFilter.toByteArray()
            
            // 反序列化
            val restoredFilter = BloomFilter.fromByteArray<Int>(bytes)
            
            // 验证
            assert(restoredFilter.size() == originalFilter.size()) { "大小应该一致" }
            
            // 检查所有元素
            var allMatch = true
            for (i in 1..500) {
                if (restoredFilter.mightContain(i) != originalFilter.mightContain(i)) {
                    allMatch = false
                    break
                }
            }
            assert(allMatch) { "序列化前后应该保持一致性" }
            
            println("\n  序列化测试:")
            println("    原始过滤器: $originalFilter")
            println("    恢复过滤器: $restoredFilter")
            println("    序列化大小: ${bytes.size} bytes")
            
            results.add(TestResult(testName, true, "序列化大小=${bytes.size} bytes"))
        } catch (e: Exception) {
            results.add(TestResult(testName, false, e.message ?: "Unknown error"))
        }
    }
    
    /**
     * 测试可扩展布隆过滤器
     */
    private fun testScalableBloomFilter() {
        val testName = "可扩展布隆过滤器测试"
        try {
            val filter = ScalableBloomFilter<String>(initialCapacity = 100, fpp = 0.01)
            
            // 添加超过初始容量的元素
            for (i in 1..1000) {
                filter.add("item_$i")
            }
            
            assert(filter.size() == 1000) { "大小应为1000" }
            assert(filter.filterCount() > 1) { "应该创建了多个过滤器" }
            
            // 检查元素
            var allContained = true
            for (i in 1..1000) {
                if (!filter.mightContain("item_$i")) {
                    allContained = false
                    break
                }
            }
            assert(allContained) { "所有添加的元素应该能被找到" }
            
            println("\n  可扩展布隆过滤器:")
            println("    过滤器数量: ${filter.filterCount()}")
            println("    总大小: ${filter.size()}")
            println("    内存使用: ${filter.memoryUsage()} bytes")
            
            results.add(TestResult(testName, true, "过滤器数量=${filter.filterCount()}"))
        } catch (e: Exception) {
            results.add(TestResult(testName, false, e.message ?: "Unknown error"))
        }
    }
    
    /**
     * 测试计数布隆过滤器
     */
    private fun testCountingBloomFilter() {
        val testName = "计数布隆过滤器测试"
        try {
            val filter = CountingBloomFilter<String>(1000, 0.01)
            
            // 添加元素
            filter.add("hello")
            filter.add("world")
            
            assert(filter.mightContain("hello")) { "hello应该存在" }
            assert(filter.mightContain("world")) { "world应该存在" }
            assert(filter.size() == 2) { "大小应为2" }
            
            // 删除元素
            assert(filter.remove("hello")) { "删除hello应该成功" }
            assert(!filter.mightContain("hello")) { "删除后hello不应该存在" }
            assert(filter.mightContain("world")) { "world应该仍然存在" }
            
            // 删除不存在的元素
            assert(!filter.remove("nonexistent")) { "删除不存在的元素应返回false" }
            
            results.add(TestResult(testName, true))
        } catch (e: Exception) {
            results.add(TestResult(testName, false, e.message ?: "Unknown error"))
        }
    }
    
    /**
     * 性能测试
     */
    private fun testPerformance() {
        val testName = "性能测试"
        try {
            val filter = BloomFilter<String>(100000, 0.01)
            val testSize = 100000
            
            // 插入性能
            val insertTime = measureNanoTime {
                for (i in 1..testSize) {
                    filter.add("item_$i")
                }
            }
            
            // 查询性能
            val queryTime = measureNanoTime {
                for (i in 1..testSize) {
                    filter.mightContain("item_$i")
                }
            }
            
            val insertNs = insertTime / testSize
            val queryNs = queryTime / testSize
            
            println("\n  性能测试 ($testSize 次操作):")
            println("    平均插入时间: ${insertNs} ns/次")
            println("    平均查询时间: ${queryNs} ns/次")
            println("    总插入时间: ${insertTime / 1_000_000} ms")
            println("    总查询时间: ${queryTime / 1_000_000} ms")
            
            // 性能断言：单次操作应在合理范围内
            assert(insertNs < 10000) { "插入操作应该很快" }
            assert(queryNs < 10000) { "查询操作应该很快" }
            
            results.add(TestResult(testName, true, "插入=${insertNs}ns, 查询=${queryNs}ns"))
        } catch (e: Exception) {
            results.add(TestResult(testName, false, e.message ?: "Unknown error"))
        }
    }
    
    /**
     * 内存效率测试
     */
    private fun testMemoryEfficiency() {
        val testName = "内存效率测试"
        try {
            val expectedInsertions = 100000
            val filter = BloomFilter<String>(expectedInsertions, 0.01)
            
            val memoryBytes = filter.memoryUsage()
            val bitsPerElement = filter.bitArraySize().toDouble() / expectedInsertions
            
            println("\n  内存效率:")
            println("    预期插入数: $expectedInsertions")
            println("    位数组大小: ${filter.bitArraySize()} bits")
            println("    内存使用: ${memoryBytes / 1024} KB")
            println("    每元素位数: ${String.format("%.2f", bitsPerElement)}")
            println("    哈希函数数: ${filter.hashFunctionCount()}")
            
            // 与HashSet对比（粗略估计）
            val hashSetEstimateBytes = expectedInsertions * 100L // 假设每个元素100字节
            val savings = (1 - memoryBytes.toDouble() / hashSetEstimateBytes) * 100
            
            println("    相比HashSet节省: ${String.format("%.1f", savings)}%")
            
            results.add(TestResult(testName, true, "内存=${memoryBytes/1024}KB, 节省${String.format("%.1f", savings)}%"))
        } catch (e: Exception) {
            results.add(TestResult(testName, false, e.message ?: "Unknown error"))
        }
    }
    
    private fun printResults() {
        println("\n========================================")
        println("  测试结果汇总")
        println("========================================")
        
        val passed = results.count { it.passed }
        val total = results.size
        
        results.forEach { result ->
            val status = if (result.passed) "✓ PASS" else "✗ FAIL"
            println("  $status - ${result.name}")
            if (result.message.isNotEmpty()) {
                println("         ${result.message}")
            }
            if (!result.passed) {
                println("         错误: ${result.message}")
            }
        }
        
        println("\n  总计: $passed/$total 通过")
        println("========================================\n")
    }
}

/**
 * 主函数入口
 */
fun main() {
    BloomFilterTest.runAllTests()
}

File: C#/bloom_filter_utils/BloomFilterTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using AllToolkit.BloomFilter;

namespace AllToolkit.Tests
{
    /// <summary>
    /// BloomFilter 单元测试
    /// </summary>
    class BloomFilterTest
    {
        private static int passed = 0;
        private static int failed = 0;
        private static readonly Random random = new Random(42);

        static void Main(string[] args)
        {
            Console.WriteLine("Running BloomFilter tests...\n");

            TestBasicOperations();
            TestFalsePositiveRate();
            TestCloneAndUnion();
            TestSerialization();
            TestScalableBloomFilter();
            TestCountingBloomFilter();
            TestEdgeCases();
            TestPerformance();

            Console.WriteLine("\n========================================");
            Console.WriteLine($"Results: {passed} passed, {failed} failed");
            Console.WriteLine("========================================");

            Environment.Exit(failed > 0 ? 1 : 0);
        }

        static void Assert(bool condition, string testName)
        {
            if (condition)
            {
                Console.WriteLine($"  ✓ {testName}");
                passed++;
            }
            else
            {
                Console.WriteLine($"  ✗ {testName}");
                failed++;
            }
        }

        static void TestBasicOperations()
        {
            Console.WriteLine("\nBasic Operations Tests:");

            // 创建布隆过滤器
            var filter = new BloomFilter<string>(1000, 0.01);
            Assert(filter.ExpectedItems == 1000, "Expected items set correctly");
            Assert(filter.Count == 0, "Initial count is 0");
            Assert(filter.BitSize > 0, "Bit size calculated");
            Assert(filter.HashFunctionCount > 0, "Hash function count calculated");

            // 添加元素
            filter.Add("hello");
            filter.Add("world");
            Assert(filter.Count == 2, "Count after adding elements");

            // 检查存在
            Assert(filter.MightContain("hello"), "Contains 'hello'");
            Assert(filter.MightContain("world"), "Contains 'world'");
            Assert(!filter.MightContain("notexist"), "Does not contain 'notexist'");

            // Contains 别名测试
            Assert(filter.Contains("hello"), "Contains alias works");

            // 批量添加
            var items = new List<string> { "one", "two", "three", "four", "five" };
            filter.AddRange(items);
            Assert(filter.Count == 7, "Count after AddRange");
            Assert(filter.MightContain("one"), "Contains 'one' after AddRange");

            // 清空
            filter.Clear();
            Assert(filter.Count == 0, "Count after Clear");
            Assert(!filter.MightContain("hello"), "Does not contain after Clear");
        }

        static void TestFalsePositiveRate()
        {
            Console.WriteLine("\nFalse Positive Rate Tests:");

            int expectedItems = 10000;
            double targetFPR = 0.01;
            var filter = new BloomFilter<string>(expectedItems, targetFPR);

            // 添加元素
            var addedItems = new HashSet<string>();
            for (int i = 0; i < expectedItems; i++)
            {
                string item = $"item_{i}";
                filter.Add(item);
                addedItems.Add(item);
            }

            Assert(filter.Count == expectedItems, "All items added");

            // 测试假阳性率
            int falsePositives = 0;
            int testCount = 100000;
            for (int i = 0; i < testCount; i++)
            {
                string testItem = $"test_{i}";
                if (!addedItems.Contains(testItem) && filter.MightContain(testItem))
                {
                    falsePositives++;
                }
            }

            double actualFPR = (double)falsePositives / testCount;
            Console.WriteLine($"    Actual FPR: {actualFPR:P4} (target: {targetFPR:P2})");
            
            // 假阳性率应该接近目标值（允许一些偏差）
            Assert(actualFPR < targetFPR * 3, $"False positive rate within bounds ({actualFPR:P4} < {targetFPR * 3:P2})");

            // 测试估计的假阳性概率
            double estimatedFPR = filter.EstimatedFalsePositiveProbability;
            Console.WriteLine($"    Estimated FPR: {estimatedFPR:P4}");
            Assert(estimatedFPR >= 0 && estimatedFPR < 1, "Estimated FPR in valid range");

            // 测试统计信息
            var stats = filter.GetStats();
            Assert(stats.ItemsAdded == expectedItems, "Stats: ItemsAdded correct");
            Assert(stats.BitSize == filter.BitSize, "Stats: BitSize correct");
            Console.WriteLine($"    Stats: {stats}");
        }

        static void TestCloneAndUnion()
        {
            Console.WriteLine("\nClone and Union Tests:");

            var filter1 = new BloomFilter<int>(1000, 0.01);
            for (int i = 0; i < 100; i++)
            {
                filter1.Add(i);
            }

            // 克隆测试
            var cloned = (BloomFilter<int>)filter1.Clone();
            Assert(cloned.Count == filter1.Count, "Clone has same count");
            Assert(cloned.BitSize == filter1.BitSize, "Clone has same bit size");

            // 验证克隆后的内容
            for (int i = 0; i < 100; i++)
            {
                Assert(cloned.MightContain(i), $"Clone contains {i}");
            }

            // Union 测试
            var filter2 = new BloomFilter<int>(1000, 0.01);
            for (int i = 100; i < 200; i++)
            {
                filter2.Add(i);
            }

            filter1.UnionWith(filter2);

            // 验证 union 后包含两个集合的元素
            for (int i = 0; i < 200; i++)
            {
                Assert(filter1.MightContain(i), $"Union contains {i}");
            }

            // Intersect 测试
            var filter3 = new BloomFilter<int>(1000, 0.01);
            var filter4 = new BloomFilter<int>(1000, 0.01);

            for (int i = 0; i < 100; i++) filter3.Add(i);
            for (int i = 50; i < 150; i++) filter4.Add(i);

            filter3.IntersectWith(filter4);

            // 验证交集
            int intersectionCount = 0;
            for (int i = 0; i < 150; i++)
            {
                if (filter3.MightContain(i))
                {
                    intersectionCount++;
                }
            }
            Console.WriteLine($"    Intersection found ~{intersectionCount} items");
            Assert(intersectionCount > 0, "Intersection contains some items");
        }

        static void TestSerialization()
        {
            Console.WriteLine("\nSerialization Tests:");

            var filter = new BloomFilter<string>(1000, 0.01);
            filter.Add("test1");
            filter.Add("test2");
            filter.Add("test3");

            // 序列化
            byte[] data = filter.ToByteArray();
            Assert(data != null && data.Length > 0, "Serialized to byte array");
            Console.WriteLine($"    Serialized size: {data.Length} bytes");

            // 反序列化
            var restored = BloomFilter<string>.FromByteArray(data, 1000, 0.01);
            Assert(restored.MightContain("test1"), "Restored contains 'test1'");
            Assert(restored.MightContain("test2"), "Restored contains 'test2'");
            Assert(restored.MightContain("test3"), "Restored contains 'test3'");
            Assert(!restored.MightContain("notexist"), "Restored does not contain 'notexist'");

            // GetSerializedSize
            int expectedSize = filter.GetSerializedSize();
            Assert(data.Length == expectedSize, "Serialized size matches GetSerializedSize");
        }

        static void TestScalableBloomFilter()
        {
            Console.WriteLine("\nScalable Bloom Filter Tests:");

            var filter = new ScalableBloomFilter<int>(100, 0.01);

            // 添加大量元素，超过初始容量
            int itemCount = 1000;
            for (int i = 0; i < itemCount; i++)
            {
                filter.Add(i);
            }

            Assert(filter.Count == itemCount, $"Count is {itemCount}");
            Assert(filter.FilterCount > 1, $"Multiple filters created ({filter.FilterCount})");

            // 验证所有元素都可被检测
            int foundCount = 0;
            for (int i = 0; i < itemCount; i++)
            {
                if (filter.MightContain(i)) foundCount++;
            }
            Assert(foundCount == itemCount, "All items found in scalable filter");

            // 获取所有统计
            var allStats = filter.GetAllStats();
            Console.WriteLine($"    Filter layers: {allStats.Count}");
            foreach (var stats in allStats)
            {
                Console.WriteLine($"      {stats}");
            }
        }

        static void TestCountingBloomFilter()
        {
            Console.WriteLine("\nCounting Bloom Filter Tests:");

            var filter = new CountingBloomFilter<string>(1000, 0.01);

            // 添加元素
            filter.Add("item1");
            filter.Add("item2");
            filter.Add("item3");
            Assert(filter.Count == 3, "Count after adding elements");

            // 检查存在
            Assert(filter.MightContain("item1"), "Contains 'item1'");
            Assert(filter.MightContain("item2"), "Contains 'item2'");
            Assert(!filter.MightContain("notexist"), "Does not contain 'notexist'");

            // 删除元素
            bool removed = filter.Remove("item1");
            Assert(removed, "Remove 'item1' succeeded");
            Assert(filter.Count == 2, "Count after remove");

            // 删除后检查
            // 注意：由于计数布隆过滤器的特性，删除后可能仍显示存在（假阳性）
            // 但我们可以验证可以删除存在的元素

            // 删除不存在的元素应该失败
            removed = filter.Remove("notexist");
            Assert(!removed, "Remove non-existent item fails");

            // 清空
            filter.Clear();
            Assert(filter.Count == 0, "Count after Clear");

            // 测试假阳性概率估计
            var filter2 = new CountingBloomFilter<string>(1000, 0.01);
            for (int i = 0; i < 500; i++)
            {
                filter2.Add($"item_{i}");
            }
            double fpr = filter2.EstimatedFalsePositiveProbability;
            Console.WriteLine($"    Estimated FPR: {fpr:P4}");
            Assert(fpr >= 0 && fpr < 1, "Counting filter FPR in valid range");
        }

        static void TestEdgeCases()
        {
            Console.WriteLine("\nEdge Cases Tests:");

            // 小容量过滤器
            var smallFilter = new BloomFilter<string>(10, 0.1);
            smallFilter.Add("a");
            smallFilter.Add("b");
            Assert(smallFilter.Count == 2, "Small filter works");
            Assert(smallFilter.MightContain("a"), "Small filter contains 'a'");

            // 高假阳性率容忍
            var highFPFilter = new BloomFilter<int>(100, 0.5);
            for (int i = 0; i < 50; i++)
            {
                highFPFilter.Add(i);
            }
            Assert(highFPFilter.Count == 50, "High FP filter works");

            // 低假阳性率
            var lowFPFilter = new BloomFilter<int>(100, 0.001);
            for (int i = 0; i < 50; i++)
            {
                lowFPFilter.Add(i);
            }
            Assert(lowFPFilter.Count == 50, "Low FP filter works");

            // 空过滤器操作
            var emptyFilter = new BloomFilter<string>(100);
            Assert(!emptyFilter.MightContain("anything"), "Empty filter contains nothing");
            Assert(emptyFilter.Count == 0, "Empty filter count is 0");

            // 类型测试
            var intFilter = new BloomFilter<int>(100);
            intFilter.Add(42);
            Assert(intFilter.MightContain(42), "Int filter contains 42");
            Assert(!intFilter.MightContain(0), "Int filter does not contain 0");

            var guidFilter = new BloomFilter<Guid>(100);
            var guid = Guid.NewGuid();
            guidFilter.Add(guid);
            Assert(guidFilter.MightContain(guid), "Guid filter contains added guid");
            Assert(!guidFilter.MightContain(Guid.NewGuid()), "Guid filter does not contain random guid");

            // 静态方法测试
            int optimalBits = BloomFilter<string>.OptimalBitSize(1000, 0.01);
            Assert(optimalBits > 0, "OptimalBitSize returns positive value");
            Console.WriteLine($"    Optimal bits for 1000 items, 1% FP: {optimalBits}");

            int optimalHashes = BloomFilter<string>.OptimalHashFunctionCount(optimalBits, 1000);
            Assert(optimalHashes > 0, "OptimalHashFunctionCount returns positive value");
            Console.WriteLine($"    Optimal hash functions: {optimalHashes}");
        }

        static void TestPerformance()
        {
            Console.WriteLine("\nPerformance Tests:");

            int itemCount = 100000;
            double targetFPR = 0.01;

            // 创建和添加性能
            var sw = System.Diagnostics.Stopwatch.StartNew();
            var filter = new BloomFilter<string>(itemCount, targetFPR);
            for (int i = 0; i < itemCount; i++)
            {
                filter.Add($"item_{i}");
            }
            sw.Stop();
            Console.WriteLine($"    Add {itemCount:N0} items: {sw.ElapsedMilliseconds} ms");
            Assert(filter.Count == itemCount, "All items added in performance test");

            // 查询性能
            sw.Restart();
            int foundCount = 0;
            for (int i = 0; i < itemCount; i++)
            {
                if (filter.MightContain($"item_{i}")) foundCount++;
            }
            sw.Stop();
            Console.WriteLine($"    Query {itemCount:N0} existing items: {sw.ElapsedMilliseconds} ms");
            Assert(foundCount == itemCount, "All items found in performance test");

            // 不存在元素的查询性能
            sw.Restart();
            for (int i = 0; i < itemCount; i++)
            {
                filter.MightContain($"notexist_{i}");
            }
            sw.Stop();
            Console.WriteLine($"    Query {itemCount:N0} non-existing items: {sw.ElapsedMilliseconds} ms");

            // 内存使用
            long memoryBytes = filter.GetSerializedSize();
            Console.WriteLine($"    Memory usage: {memoryBytes:N0} bytes ({memoryBytes / 1024.0:N2} KB)");
            Console.WriteLine($"    Bits per item: {(double)filter.BitSize / itemCount:N2}");
        }
    }
}